### DISCUSSION TITLE
"Cannot use import statement outside a module." when i set tsNode=true

### DISCUSSION THREAD

**Discussion by yangyyou** [mikro-orm/mikro-orm#4738] [General]
**Describe**
I user mikro-orm in a Nestjs project,  set tsNode= true. There occurred a problem
```Nest] 53168  - 2023/09/24 10:18:35   ERROR [ExceptionHandler] Cannot use import statement outside a module
/Users/Code/nestjs/book-admin/src/menu/entities/menu.entity.ts:1
import {
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1178:20)
    at Module._compile (node:internal/modules/cjs/loader:1220:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Function.Module._load (node:internal/modules/cjs/loader:960:12)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Function.requireFrom (/Users/Code/nestjs/book-admin/node_modules/@mikro-orm/core/utils/Utils.js:759:70)
    at Function.options.dynamicImportProvider.options.dynamicImportProvider [as dynamicImportProvider] (/Users/Code/nestjs/book-admin/node_modules/@mikro-orm/core/utils/ConfigurationLoader.js:226:34)
```

**mikro-orm.config.ts**
```typescript
const config: MikroOrmModuleOptions = {
  tsNode: true,
  entities: [__dirname + '/../dist/**/*.entity.js'],
  entitiesTs: [__dirname + '/../src/**/*.entity.ts'],
  baseDir: process.cwd(),
  autoLoadEntities: true,
  dbName: 'book_admin',
  type: 'mysql',
  host: 'localhost',
  port: 3306,
  user: 'root',
  password: 'root',
  debug: true,
  logger: logger.log.bind(logger),
  highlighter: new SqlHighlighter(),
  metadataProvider: TsMorphMetadataProvider
};
```

**tsconfig setting**
```JSON
{
  "compilerOptions": {
    "module": "commonjs",
    "declaration": true,
    "removeComments": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "allowSyntheticDefaultImports": true,
    "target": "ES2021",
    "sourceMap": true,
    "outDir": "./dist",
    "baseUrl": "./",
    "incremental": true,
    "skipLibCheck": true,
    "strictNullChecks": false,
    "noImplicitAny": false,
    "strictBindCallApply": false,
    "forceConsistentCasingInFileNames": false,
    "noFallthroughCasesInSwitch": false,
    "esModuleInterop": true
  }
}
```

**Version**
"@mikro-orm/core": "^5.8.2",
    "@mikro-orm/mysql": "^5.8.2",
    "@mikro-orm/nestjs": "^5.2.1",
    "@nestjs/common": "^10.0.0",
    "@nestjs/core": "^10.0.0",
    "ts-node": "^10.9.1",

**Comment by B4nan** [comment]
It's up to you to bring TS support to your project (if you want it), e.g. by executing it via `ts-node` instead of `node` - the ORM won't register ts-node for you, that works only for the CLI. `tsNode: true` in the config only tells the ORM that it should use TS files, expecting that you already brought the TS support so they can be imported, you are not supposed to enable this flag yourself (outside of test environments), the value is set automatically based on whether the TS support is detected or not. You definitely don't want to have `tsNode: true` in production.

You can take a look at this project, it has everything set up correctly: 

https://github.com/mikro-orm/nestjs-realworld-example-app

Not sure if defaults changed in nest templates, I know they used to have ts-node in place for the dev mode, but now they might have switched to compiling things on the fly via webpack, I know they have this webpack flag in their config - that literally means there is no dev mode for the ORM either, and you need to compile your entities and configs (and use paths to that).

---

Btw I also see that you try to use both folder-based discovery (`entities/entitiesTs` options) with auto-loading (`autoLoadEntities`) - you should only need one or the other.
